step-automation-packages/step-automation-packages-manager/src/main/java/step/automation/packages/AutomationPackageManager.java

```diff
@@ -259,35 +266,89 @@ public PackageUpdateResult createOrUpdateAutomationPackage(boolean allowUpdate,
+                ObjectId result = updateAutomationPackage(oldPackage, newPackage, staging, enricherForIncludedEntities, true);
+                return new AutomationPackageUpdateResult(AutomationPackageUpdateStatus.UPDATED, result);
+            } else {
+                // async update
+                newPackage.setStatus(AutomationPackageStatus.DELAYED_UPDATE);
+                automationPackageAccessor.save(newPackage);
+                AutomationPackage finalNewPackage = newPackage;
+                delayedUpdateExecutor.submit(() -> {
+                    try {
+                        updateAutomationPackage(oldPackage, finalNewPackage, staging, enricherForIncludedEntities, false);
+                    } catch (Exception e) {
+                        handleExceptionOnPackageUpdate(finalNewPackage);
+                        log.error("Exception on delayed AP update", e);
+                    }
+                });
+                return new AutomationPackageUpdateResult(AutomationPackageUpdateStatus.UPDATE_DELAYED, newPackage.getId());
+            }
+        } catch (Exception ex) {
+            handleExceptionOnPackageUpdate(newPackage);
+            throw ex;
+        }
+    }
+
+    private ObjectId updateAutomationPackage(AutomationPackage oldPackage, AutomationPackage newPackage,
+                                             Staging staging, ObjectEnricher enricherForIncludedEntities,
+                                             boolean alreadyLocked) {
+        try {
+            //If not already locked (i.e. was not able to acquire an immediate write lock)
+            if (!alreadyLocked) {
+                getWriteLock(newPackage);
+            }
             // delete old package entities
             if (oldPackage != null) {
                 deleteAutomationPackageEntities(oldPackage);
             }
-
             // persist all staged entities
             persistStagedEntities(staging, enricherForIncludedEntities);
 
-            // save automation package metadata
+            //Clear status in case previous async update failed
+            newPackage.setStatus(null);
             ObjectId result = automationPackageAccessor.save(newPackage).getId();
 
             logAfterSave(staging, oldPackage, newPackage);
-            return new PackageUpdateResult(oldPackage == null ? PackageUpdateStatus.CREATED : PackageUpdateStatus.UPDATED, result);
-        } catch (Exception ex) {
-            // cleanup created resources
-            try {
-                if (newPackage != null) {
-                    List<Resource> resources = resourceManager.findManyByCriteria(Map.of("customFields." + AutomationPackageEntity.AUTOMATION_PACKAGE_ID, newPackage.getId().toString()));
-                    for (Resource resource : resources) {
-                        resourceManager.deleteResource(resource.getId().toString());
-                    }
+            return result;
+        } finally {
+            releaseWriteLock(newPackage);
```

Comment:
@david-stephan 
Shouldn't we also release the 'DELAYED_UPDATE' status in finally block?

Reply:
correct, changed it